packages/mobile/src/hooks/useNavigation.ts

```diff
@@ -1,95 +1,73 @@
+export type ContextualParams = {
+  fromNotifications?: boolean
+}
 
-export type ContextualParams = { fromNotifications?: boolean }
+export type ContextualizedParamList<ParamList extends ParamListBase> = {
+  [K in keyof ParamList]: ParamList[K] & ContextualParams
+}
 
-type UseNavigationConfig<
+type PerformNavigationConfig<
   ParamList extends ParamListBase,
-  RouteName extends keyof ParamList
+  RouteName extends keyof ParamList = keyof ParamList
 > = [screen: RouteName, params?: ParamList[RouteName] & ContextualParams]
 
-type UseNavigationOptions = {
-  customNativeNavigation?: DrawerNavigationHelpers
+type UseNavigationOptions<NavigationProp extends RNNavigationProp<any>> = {
+  customNavigation?: NavigationProp
 }
 
-export const useNavigation = <
-  ParamList extends ParamListBase = AppTabScreenParamList
->(
-  options?: UseNavigationOptions
-) => {
-  const { navigation: defaultNativeNavigation } = useContext(
-    AppTabNavigationContext
-  )
+/**
+ * Custom wrapper around react-navigation `useNavigation`
+ *
+ * Features:
+ * - Prevent duplicate navigation pushes
+ * - Apply contextual params to all routes
+ */
+export function useNavigation<
+  ParamList extends ParamListBase,
+  NavigationProp extends RNNavigationProp<ParamListBase> = NativeStackNavigationProp<ParamList>
+>(options?: UseNavigationOptions<NavigationProp>): NavigationProp {
+  const defaultNavigation = useNativeNavigation<NavigationProp>()
 
-  const nativeNavigation =
-    options?.customNativeNavigation || defaultNativeNavigation
+  const lastNavAction = useRef<PerformNavigationConfig<ParamList>>()
 
-  const performNavigation = useCallback(
-    (method) =>
-      <RouteName extends keyof ParamList>(
-        ...config: UseNavigationConfig<ParamList, RouteName>
-      ) => {
-        const [screen, params] = config
-        method(screen, params)
-      },
-    []
-  )
+  const navigation: NavigationProp =
+    options?.customNavigation ?? defaultNavigation
 
+  // Prevent duplicate pushes by de-duping
+  // navigation actions
   const performCustomPush = useCallback(
-    <RouteName extends keyof ParamList>(
-      ...config: UseNavigationConfig<ParamList, RouteName>
-    ) => {
-      const [screen, params] = config
-
-      const customPushAction = (state: StackNavigationState<ParamList>) => {
-        const lastRoute = state.routes[state.routes.length - 1]
-        if (
-          (screen as string) === lastRoute.name &&
-          isEqual(params, lastRoute.params)
-        ) {
-          // react-navigation considers this a no-op
-          return CommonActions.navigate(lastRoute)
-        }
-        return StackActions.push(screen as string, params)
+    (...config: PerformNavigationConfig<ParamList>) => {
+      if (!isEqual(lastNavAction.current, config)) {
+        ;(navigation as NativeStackNavigationProp<any>).push(...config)
+        lastNavAction.current = config
+        setTimeout(() => {
```

Comment:
okay this is interesting, what's the idea here? imo maybe it's okay to not allow a user to push to the same page they are currently on no matter what?

Reply:
So it's to account for popping off the stack and then trying to push the previous thing again. Only the push action is stored here so the next push would get cancelled

Reply:
That one only gets called when we call it, not when swiping back. I went down the path of using an event handler but it then gets called it every instance of useNavigation

Reply:
tricky, well sg!